File: Java/Hard/CompileError/number-of-ways-to-separate-numbers_6.java
class Solution {
    public int numberOfCombinations(String num) {
        
        int left = 1, right = cells.length;
        int[][] grid = new int[row][col];
        while (left < right) {
            int mid = (left + right + 1) / 2;
            for (int i = 0; i < row; i++) {
                Arrays.fill(grid[i], 0);
            }
            for (int i = 0; i < mid; i++) {
                grid[cells[i][0] - 1][cells[i][1] - 1] = 1;
            }
            if (dfs(grid, 0, 0)) {
                left = mid;
            } else {
                right = mid - 1;
            }
        }
        return left;
    }

    private boolean dfs(int[][] grid, int i, int j) {
        int row = grid.length, col = grid[0].length;
        if (i == row - 1) {
            return true;
        }
        if (i < 0 || i >= row || j < 0 || j >= col || grid[i][j] == 1) {
            return false;
        }
        grid[i][j] = 1;
        boolean res = dfs(grid, i + 1, j) || dfs(grid, i - 1, j) || dfs(grid, i, j + 1) || dfs(grid, i, j - 1);
        grid[i][j] = 0;
        return res;
    }
}

File: C++/Hard/WrongAnswer/number-of-ways-to-separate-numbers_6.cpp
class Solution {
public:
    int numberOfCombinations(string num) {
        
        int n = num.size();
        vector<int> dp(n + 1, 0);
        vector<int> count(n + 1, 0);
        dp[0] = 1;
        count[0] = 1;
        for (int i = 1; i <= n; ++i) {
            for (int j = 0; j < i; ++j) {
                if (num[j] != '0' && stol(num.substr(j, i - j)) <= 1000000000) {
                    dp[i] = (dp[i] + dp[j]) % 1000000007;
                    count[i] = (count[i] + count[j]) % 1000000007;
                } else {
                    break;
                }
            }
        }
        return count[n];
    }
};